daml-script/export/src/main/scala/com/daml/script/export/Config.scala

```diff
@@ -0,0 +1,123 @@
+    acsBatchSize: Int,
+    outputPath: Path,
+    sdkVersion: String,
+    damlScriptLib: String,
+) extends ExportType
+
+object Config {
+  def parse(args: Array[String]): Option[Config] =
+    parser.parse(args, Empty)
+
+  private def parseLedgerOffset(s: String): LedgerOffset = LedgerOffset {
+    s match {
+      case s if s == "begin" =>
+        LedgerOffset.Value.Boundary(LedgerOffset.LedgerBoundary.LEDGER_BEGIN)
+      case s if s == "end" => LedgerOffset.Value.Boundary(LedgerOffset.LedgerBoundary.LEDGER_END)
+      case s => LedgerOffset.Value.Absolute(s)
+    }
+  }
+
+  private val parser = new scopt.OptionParser[Config]("script-export") {
+    help("help")
+      .text("Show this help message.")
+    opt[String]("host")
+      .required()
+      .action((x, c) => c.copy(ledgerHost = x))
+      .text("Daml ledger host to connect to.")
+    opt[Int]("port")
+      .required()
+      .action((x, c) => c.copy(ledgerPort = x))
+      .text("Daml ledger port to connect to.")
+    opt[String]("party")
+      .required()
+      .unbounded()
+      .action((x, c) => c.copy(parties = x :: c.parties))
+      .text("Export ledger state as seen by these parties.")
+    opt[String]("start")
+      .optional()
+      .action((x, c) => c.copy(start = parseLedgerOffset(x)))
+      .text(
+        "The transaction offset (exclusive) for the start position of the export. Optional, by default the export includes the beginning of the ledger."
+      )
+    opt[String]("end")
+      .optional()
+      .action((x, c) => c.copy(start = parseLedgerOffset(x)))
+      .text(
+        "The transaction offset (inclusive) for the end position of the export. Optional, by default the export includes the current end of the ledger."
+      )
+    cmd("script")
+      .action((_, c) => c.copy(exportType = Some(EmptyExportScript)))
+      .text("Export ledger state in Daml script format")
+      .children(
+        opt[Int]("acs-batch-size")
+          .optional()
+          .action(actionExportScript((x, c) => c.copy(acsBatchSize = x)))
+          .validate(x =>
+            if (x <= 0) { failure("ACS batch size must be greater than zero") }
+            else { success }
+          )
+          .text("Batch this many create commands into one transaction when recreating the ACS."),
+        opt[File]('o', "output")
+          .required()
+          .action(actionExportScript((x, c) => c.copy(outputPath = x.toPath)))
+          .text("Create the Daml script under this directory prefix."),
+        opt[String]("sdk-version")
+          .required()
+          .action(actionExportScript((x, c) => c.copy(sdkVersion = x)))
+          .text("Specify this Daml Connect version in the generated project."),
+      )
+    checkConfig(c =>
+      c.exportType match {
+        case None => failure("Must specify export type")
+        case Some(_) => success
+      }
+    )
+  }
+
+  private val EmptyExportScript = ExportScript(
+    outputPath = null,
```

Comment:
This is unlikely to be a cause of concern and I understand that this is hidden behind a required option, but is there something we can do to make errors more telling than a `NullPointerException`? Would it be feasible to have a "raw" version of this that contains `Option`s and have a function that validates it and turns it into something that can be used safely?

Reply:
Definitely feasible, I tend to go for the `null/""/…` approach personally mainly because I’m lazy and I’m angry at scopt for making me implement my own validation logic (Haskell’s optparse-applicative is much nicer). That said, that’s a pretty poor excuse so definitely open to reconsider this. But let’s leave it out of this PR, this is just reshuffling.